qutip/core/qobj.py

```diff
@@ -473,12 +473,31 @@ def __rsub__(self, other):
 
     @_tidyup
     def __mul__(self, other):
-        if not isinstance(other, numbers.Number):
+        """If other is a Qobj, we dispatch to __matmul__. If not, we
+        check that other is a valid complex scalar, i.e., we can do
+        complex(other). Otherwise, we return NotImplemented."""
+
+        if isinstance(other, Qobj):
             return self.__matmul__(other)
-        multiplier = complex(other)
-        isherm = (self._isherm and multiplier.imag == 0) or None
-        isunitary = (self._isunitary and abs(multiplier) == 1) or None
-        return Qobj(_data.mul(self._data, multiplier),
+
+        # We send other to mul instead of complex(other) to be more flexible. 
+        # The dispatcher can then decide how to handle other and return 
+        # TypeError if it does not know what to do with the type of other.
+        try:
+            out = _data.mul(self._data, other)
+        except TypeError:
+            return NotImplemented
+
+        # Infer isherm and isunitary if possible
+        try:
+            multiplier = complex(other)
+            isherm = (self._isherm and multiplier.imag == 0) or None
+            isunitary = (self._isunitary and abs(multiplier) == 1) or None
```

Comment:
```suggestion
            isunitary = (abs(multiplier) == 1) if self._isunitary or None
```
If `self` is unitary, then `multiplier * self` is unitary iff `abs(multiplier) == 1`.

Unfortunately we can't do something equally nice for isherm because if `self` is the zero matrix, then `self * anything` remains hermitian.

This suggestion is a bit unrelated to this PR though, but maybe it is nice to include it here anyway.

Reply:
I think the syntax of this statement is wrong. Did you mean ` isunitary = (abs(multiplier) == 1) if self._isunitary else None`?

Reply:
I did! Apologies -- I rewrote the suggestion a few times while thinking about things.

Reply:
I included the suggested change.